sys/initrd.c

```diff
@@ -0,0 +1,279 @@
+  uid_t c_uid;
+  gid_t c_gid;
+  dev_t c_rdev;
+  off_t c_size;
+  time_t c_mtime;
+
+  const char *c_path; /* contains exact path to file as archived from cpio */
+  const char *c_name; /* contains name of file */
+  void *c_data;
+} cpio_node_t;
+
+static MALLOC_DEFINE(mp, "initial ramdisk memory pool");
+typedef TAILQ_HEAD(, cpio_node) cpio_list_t;
+
+static vm_addr_t rd_start;
+static size_t rd_size;
+static cpio_list_t initrd_head;
+static vnodeops_t initrd_ops;
+
+extern char *kenv_get(const char *key);
+
+static void cpio_node_dump(cpio_node_t *cn) {
+  log("entry '%s': {dev: %ld, ino: %lu, mode: %d, nlink: %d, "
+      "uid: %d, gid: %d, rdev: %ld, size: %ld, mtime: %lu}",
+      cn->c_path, cn->c_dev, cn->c_ino, cn->c_mode, cn->c_nlink, cn->c_uid,
+      cn->c_gid, cn->c_rdev, cn->c_size, cn->c_mtime);
+}
+
+static void read_bytes(char **tape, void *ptr, size_t bytes) {
+  memcpy(ptr, *tape, bytes);
+  *tape += bytes;
+}
+
+static void skip_bytes(char **tape, size_t bytes) {
+  *tape = align(*tape + bytes, 4);
+}
+
+#define MKDEV(major, minor) (((major & 0xff) << 8) | (minor & 0xff))
+
+static bool read_cpio_header(char **tape, cpio_node_t *cpio) {
+  cpio_new_hdr_t hdr;
+
+  read_bytes(tape, &hdr, sizeof(hdr));
+
+  uint16_t c_magic = strntoul(hdr.c_magic, 6, NULL, 8);
+
+  if (c_magic != CPIO_NMAGIC && c_magic != CPIO_NCMAGIC) {
+    log("wrong magic number: %o", c_magic);
+    return false;
+  }
+
+  uint32_t c_ino = strntoul(hdr.c_ino, 8, NULL, 16);
+  uint32_t c_mode = strntoul(hdr.c_mode, 8, NULL, 16);
+  uint32_t c_uid = strntoul(hdr.c_uid, 8, NULL, 16);
+  uint32_t c_gid = strntoul(hdr.c_gid, 8, NULL, 16);
+  uint32_t c_nlink = strntoul(hdr.c_nlink, 8, NULL, 16);
+  uint32_t c_mtime = strntoul(hdr.c_mtime, 8, NULL, 16);
+  uint32_t c_filesize = strntoul(hdr.c_filesize, 8, NULL, 16);
+  uint32_t c_maj = strntoul(hdr.c_maj, 8, NULL, 16);
+  uint32_t c_min = strntoul(hdr.c_min, 8, NULL, 16);
+  uint32_t c_rmaj = strntoul(hdr.c_rmaj, 8, NULL, 16);
+  uint32_t c_rmin = strntoul(hdr.c_rmin, 8, NULL, 16);
+  uint32_t c_namesize = strntoul(hdr.c_namesize, 8, NULL, 16);
+
+  cpio->c_dev = MKDEV(c_maj, c_min);
+  cpio->c_ino = c_ino;
+  cpio->c_mode = c_mode;
+  cpio->c_nlink = c_nlink;
+  cpio->c_uid = c_uid;
+  cpio->c_gid = c_gid;
+  cpio->c_rdev = MKDEV(c_rmaj, c_rmin);
+  cpio->c_size = c_filesize;
+  cpio->c_mtime = c_mtime;
+
+  cpio->c_path = *tape;
+  skip_bytes(tape, c_namesize);
+  cpio->c_data = *tape;
+  skip_bytes(tape, c_filesize);
+
+  return true;
+}
+
+static void read_cpio_archive() {
+  char *tape = (char *)rd_start;
+
+  while (true) {
+    cpio_node_t *node = kmalloc(mp, sizeof(cpio_node_t), M_ZERO);
+    TAILQ_INIT(&node->c_children);
+    if (!read_cpio_header(&tape, node) ||
+        strcmp(node->c_path, CPIO_TRAILER) == 0) {
+      kfree(mp, node);
+      break;
+    }
+    TAILQ_INSERT_TAIL(&initrd_head, node, c_list);
+  }
+}
+
+static bool is_direct_descendant(const char *A, const char *B) {
```

Comment:
Hmm... could you use [string](http://netbsd.gw.com/cgi-bin/man-cgi?string) functions from standard library to perform this task - there's a lot of string processing magic below that is difficult to read.

If one you'd like to use is not in the kernel please feel free to take one from NetBSD / FreeBSD - I can help you with that.